tensorpipe/channel/cuda/context.cc

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/cuda/context.h>
+
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <limits>
+#include <list>
+
+#include <tensorpipe/channel/cuda/channel.h>
+#include <tensorpipe/channel/error.h>
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/channel/registry.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error_macros.h>
+#include <tensorpipe/common/optional.h>
+#include <tensorpipe/common/queue.h>
+#include <tensorpipe/common/system.h>
+#include <tensorpipe/proto/channel/cuda.pb.h>
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda {
+
+namespace {
+
+std::string generateDomainDescriptor() {
+  std::ostringstream oss;
+  auto bootID = getBootID();
+  TP_THROW_ASSERT_IF(!bootID) << "Unable to read boot_id";
+  oss << bootID.value();
+  return oss.str();
```

Comment:
The CUDA docs state:
> cudaIpcOpenMemHandle can open handles to devices that may not be visible in the process calling the API.

> In addition, we still need a change in the generic channel context interface that allows a channel to mark itself as "unavailable": we need that in case we don't detect P2P links between the devices.

Agreed.

Reply:
Yes, you're right that `cudaIpcOpenMemHandle` can open handles from non-visible devices. However, we will need each channel context to check that P2P is supported between all pairs of devices that are visible to it, and mark itself as disabled if it is not. This however implies that we must also require the two sides of the channel to have the same set of visible devices.

To see why, suppose the machine has 2 GPUs, which do _not_ have IPC between them, and each endpoint of the channel only sees one of the GPUs. Each endpoint thinks there is P2P between all pairs of GPUs it can see (because, well, there is no such pair). They are both on the same machine, so the boot ID would match. And yet, they should not be allowed to use the channel.

Reply:
> However, we will need each channel context to check that P2P is supported between all pairs of devices that are visible to it, and mark itself as disabled if it is not. This however implies that we must also require the two sides of the channel to have the same set of visible devices.

 [This presentation from NVidia]() says the following, so I'm under the impression we can rely on that behavior when P2P is not enabled/available between two devices:
> Note that this [cudaMemcpy] will transparently fall back to a normal copy
through the host if P2P is not available

> To see why, suppose the machine has 2 GPUs, which do not have IPC between them, and each endpoint of the channel only sees one of the GPUs. 

By IPC, do you mean P2P? My understanding is that all we need is universal addressing to enable IPC and then `cudaMemcpy` will use P2P when available, and fall back to a copy through the host otherwise.

Reply:
Ok, I see. Do we want to let CUDA do a non-P2P memory copy on its own? Or do we want this channel to refuse to do non-P2P memcpys so that we can do them on our own? I am wondering how fast CUDA would be: for example, would it beat a multiplexed CMA channel?

Reply:
As discussed, let's let CUDA figure it out for now, and add an optimized channel for non-P2P copies later if needed.